corehq/apps/registration/forms.py

```diff
@@ -463,11 +463,26 @@ def clean_email(self):
 
 class MobileWorkerAccountConfirmationForm(BaseUserInvitationForm):
     """
-    For Mobile Workers to confirm their accounts.
+    For Mobile Workers to confirm their accounts using Email.
     """
     pass
 
 
+class MobileWorkerAccountConfirmationBySMSForm(BaseUserInvitationForm):
+    """
+    For Mobile Workers to confirm their accounts using SMS.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Mobile workers who are confirming by SMS need not require an email to input.
+        self.fields['email'].widget = forms.HiddenInput()
+        self.fields['email'].required = False
+
+    def clean_email(self):
+        # Return email without any validation
+        return self.cleaned_data['email'].strip().lower()
```

Comment:
> Usecase here assumes that the mobile workers may not have an email address.

That makes sense. And your form, as you have written it, allows for no email address, but it also allows for the possibility of an email address, and preserves it if it exists. This seems right to me.

However, your view is wiping any existing email address. Unlike your form, it is enforcing that mobile workers who confirm by SMS cannot have an email address, even in the (unlikely) scenario that they were created/imported with email addresses. That seems contradictory to your form, and surprising behavior if they were intentionally created with email addresses.

If we know for certain that we need to enforce blank email addresses for mobile workers who confirm by SMS, then `clean_email()` can just return an empty string:

```suggestion
class MobileWorkerAccountConfirmationBySMSForm(BaseUserInvitationForm):
    """
    For Mobile Workers to confirm their accounts using SMS.
    """
    email = forms.CharField(widget=forms.HiddenInput(), required=False)

    # Email address is enforced blank for mobile workers who confirm by SMS.
    def clean_email(self):
        return ''
```

But I'm still unsure why we want to enforce that the way the view does.

Reply:
>  if they were intentionally created with email addresses.
It is specific requirement from the project to not have email address field. Plus, form used to create the worker is without email address input.

Reply:
OK cool. And I see that the functionality is only available to that project, because you are using a feature flag. :+1: Thanks for explaining.